File: src/leetcode/python/two_pointers/SortColors_75.py
class SortColors_75(object):
    def sortColors(self, nums):
        zeroPos = 0
        twoPos = len(nums) - 1
        for i in range(twoPos + 1):
            while nums[i] == 2 and i < twoPos:
                nums[i], nums[zeroPos] = nums[zeroPos], nums[i]
                zeroPos += 1
            while nums[i] == 0 and i > zeroPos:
                nums[i], nums[twoPos] = nums[twoPos], nums[i]
                twoPos -= 1

File: src/leetcode/java/two_pointers/SortColors_75.java
package leetcode.java.two_pointers;

public class SortColors_75 {
  public void sortColors(int[] nums) {
    if (nums.length == 0 || nums.length == 1) return;
    int zeroPos = 0;
    int twoPos = nums.length - 1;
    for (int i = 0; i <= twoPos; i++) {
      while (nums[i] == 2 && i < twoPos) swap(nums, twoPos--, i);
      while (nums[i] == 0 && i > zeroPos) swap(nums, zeroPos++, i);
    }
  }

  private static void swap(int[] nums, int index1, int index2) {
    int tmp = nums[index1];
    nums[index1] = nums[index2];
    nums[index2] = tmp;
  }
}
